rclcpp/multithread_executor/multithread_executor.cpp

```diff
@@ -0,0 +1,194 @@
+// limitations under the License.
+
+#include <chrono>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <iostream>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+/* For this example, we will be creating a publishing node like the one in minimal_publisher.
+ * We will have a single subscriber node running 2 threads. Each thread loops at different speeds, and
+ * just repeats what it sees from the publisher to the screen.
+ */
+
+class PublisherNode : public rclcpp::Node
+{
+public:
+  PublisherNode()
+  : Node("PublisherNode")
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello World! " + std::to_string(this->count_++);
+
+        // Extract current thread
+        auto curr_thread = string_thread_id(std::this_thread::get_id());
+
+        // Prep display message
+        auto info_message = "\n<<THREAD " + curr_thread + ">> Publishing '%s'";
+        RCLCPP_INFO(this->get_logger(), info_message, message.data.c_str());
+        this->publisher_->publish(message);
+      };
+    timer_ = this->create_wall_timer(500ms, timer_callback);
+  }
+
+private:
+  /**
+   * A small convenience function for converting a thread ID to a string
+   */
+  std::string string_thread_id(std::thread::id id)
+  {
+    auto hashed = std::hash<std::thread::id>()(id);
+    return std::to_string(hashed);
+  }
+
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+class DualThreadedNode : public rclcpp::Node
+{
+public:
+  DualThreadedNode()
+  : Node("DualThreadedNode")
+  {
+    /* These define the callback groups
+     * They don't really do much on their own, but they have to exist in order to
+     * assign callbacks to them. They're also what the executor looks for when trying to run multiple threads
+     */
+    callback_group_subscriber1_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+    callback_group_subscriber2_ = this->create_callback_group(
+      rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+
+    // Each of these callback groups is basically a thread
+    // Everything assigned to one of them gets bundled into the same thread
+    sub1_opt = rclcpp::SubscriptionOptions();
+    sub1_opt.callback_group = callback_group_subscriber1_;
+    sub2_opt = rclcpp::SubscriptionOptions();
+    sub2_opt.callback_group = callback_group_subscriber2_;
+
+    subscription1 = this->create_subscription<std_msgs::msg::String>("topic",
+        rclcpp::QoS(10),
+        // std::bind is sort of C++'s way of passing a function
+        // If you're used to function-passing, skip these comments
+        std::bind(
```

Comment:
The indentation is weird for the arguments to `std::bind`; it's not clear that they are to `std::bind` instead of to `create_subscription`.  I'll suggest a style like https://github.com/ros2/rclcpp/blob/9723576821304add8993c7e9a3057fa284c1146f/rclcpp_lifecycle/test/test_register_custom_callbacks.cpp#L141 , which will also be closer to what we would do if we ever enable the linters here.

Reply:
I was attempting to follow the linting guidelines - I thought that anything inside parenthesis was indented over 2 spaces in C++. Though I will admit it's still really confusing to read.

I've modified the comments to hopefully make things a little easier to parse. We'll see if it helps in a bit.

Reply:
I don't use VSCode, so I don't know, sorry.